core/src/services/monoiofs/core.rs

```diff
@@ -0,0 +1,230 @@
+        // Since users use monoiofs in a context of tokio, all monoio
+        // operations need to be dispatched to a dedicated thread pool
+        // where a monoio runtime runs on each thread. Here we spawn
+        // these worker threads.
+        let (tx, rx) = flume::unbounded();
+        let threads = (0..worker_threads)
+            .map(move |i| {
+                let rx = rx.clone();
+                std::thread::Builder::new()
+                    .name(format!("monoiofs-worker-{i}"))
+                    .spawn(move || Self::worker_entrypoint(rx, io_uring_entries))
+                    .expect("spawn worker thread should success")
+            })
+            .collect();
+        let threads = Mutex::new(threads);
+
+        Self { root, tx, threads }
+    }
+
+    pub fn root(&self) -> &PathBuf {
+        &self.root
+    }
+
+    /// entrypoint of each worker thread, sets up monoio runtimes and channels
+    fn worker_entrypoint(rx: Receiver<TaskSpawner>, io_uring_entries: u32) {
+        let mut rt = RuntimeBuilder::<FusionDriver>::new()
+            .enable_all()
+            .with_entries(io_uring_entries)
+            .build()
+            .expect("monoio runtime initialize should success");
+        // run a infinite loop that receives TaskSpawner and calls
+        // them in a context of monoio
+        rt.block_on(async {
+            while let Ok(spawner) = rx.recv_async().await {
+                spawner();
+            }
+        })
+    }
+
+    #[allow(dead_code)]
+    /// create a TaskSpawner, send it to the thread pool and wait
+    /// for its result
+    pub async fn dispatch<F, Fut, T>(&self, f: F) -> T
+    where
+        F: FnOnce() -> Fut + 'static + Send,
+        Fut: Future<Output = T>,
+        T: 'static + Send,
+    {
+        // oneshot channel to send result back
+        let (tx, rx) = oneshot::channel();
+        self.tx
+            .send_async(Box::new(move || {
+                monoio::spawn(async move {
+                    tx.send(f().await)
+                        // discard result because it may be non-Debug and
+                        // we don't need it to appear in the panic message
+                        .map_err(|_| ())
+                        .expect("send result from worker thread should success");
+                });
+            }))
+            .await
+            .expect("send new TaskSpawner to worker thread should success");
+        match rx.await {
+            Ok(result) => result,
+            // tx is dropped without sending result, probably the worker
+            // thread has panicked.
+            Err(_) => self.propagate_worker_panic(),
+        }
+    }
+
+    /// This method always panics. It is called only when at least a
+    /// worker thread has panicked or meet a broken rx, which is
+    /// unrecoverable. It propagates worker thread's panic if there
+    /// is any and panics on normally exited thread.
+    fn propagate_worker_panic(&self) -> ! {
```

Comment:
I'm guessing `!` is not stablized yet?

Reply:
Return position never type is stable. See [Rust Reference](https://doc.rust-lang.org/1.75.0/reference/types/never.html).